[JIT] Add buffer/parameter metadata test to test_save_load.py

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49588 [JIT] Use `is_buffer` in `BufferPolicy::valid`
* **#49594 [JIT] Add buffer/parameter metadata test to test_save_load.py**

**Summary**
This commit adds a unit test to `test_save_load.py` that checks that
saving and loading a module preserves metadata about which module
attributes are parameters and buffers. The hooks that are currently used
to automatically check serialization of every function and module in the
unit tests check that the archive produced by saving and loading and
saving again are the same and that the type tags for the actual IValues
representing the module match before saving and after loading. However,
these tests do not check that buffer and parameter metadata was not
lost or destroyed during serialization.

**Test Plan**
Ran the new unit test.

Differential Revision: [D25730603](https://our.internmc.facebook.com/intern/diff/D25730603)

SplitInfinity commented:
In general, I think this a good thing to test now that we are removing the macro guard from `python_print.cpp` for buffer metadata, but I don't think future cases of forward compatibility will be able to include a unit test that doesn't break in fbcode.

I think the more general solution is to have tests that check that older version of Pytorch can deserialize archives produced by new versions of PyTorch (the opposite of backwards compatibility tests?).

xw285cornell commented on test/jit/test_save_load.py:
```diff
@@ -983,3 +983,46 @@ def forward(self) -> Optional[FooTuple]:
+        for m_p, loaded_p in zip(m.parameters(), m_loaded.parameters()):
+            self.assertEqual(m_p, loaded_p)
+
+        # Check buffers.
```
For this test case, we should see both attributes and buffer as "buffer" right?

SplitInfinity replied:
Yes, that's what TorchScript returns but my claim is that this is not correct. Eager mode output for this case is:
```
>>> class Mod(torch.nn.Module):
...   def __init__(self):
...     super().__init__()
...     self.t = torch.rand(4)
...     self.register_buffer("buffer", torch.rand(4))
... 
>>> m = Mod()
>>> list(m.buffers())
[tensor([0.0439, 0.5317, 0.5104, 0.7503])]
>>> list(m.named_buffers())
[('buffer', tensor([0.0439, 0.5317, 0.5104, 0.7503]))]
```
The diff stacked on top fixes this problem by applying the change to `BufferPolicy` that you reverted.

xw285cornell replied:
Sounds great, which diff should land first? I was thinking we should land this first (even if the behavior is not right), but if we want to land the other one that's fine with me too. Also please import the other diff so that I can try it out :)

SplitInfinity replied:
I think both need to land at once - this diff will break CI if landed by itself.

#49588 has already been imported as D25633250.

Commit:
Update on "[JIT] Add buffer/parameter metadata test to test_save_load.py"


**Summary**
This commit adds a unit test to `test_save_load.py` that checks that
saving and loading a module preserves metadata about which module
attributes are parameters and buffers. The hooks that are currently used
to automatically check serialization of every function and module in the
unit tests check that the archive produced by saving and loading and
saving again are the same and that the type tags for the actual IValues
representing the module match before saving and after loading. However,
these tests do not check that buffer and parameter metadata was not
lost or destroyed during serialization.

**Test Plan**
Ran the new unit test.

Differential Revision: [D25730603](https://our.internmc.facebook.com/intern/diff/D25730603)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add buffer/parameter metadata test to test_save_load.py"


**Summary**
This commit adds a unit test to `test_save_load.py` that checks that
saving and loading a module preserves metadata about which module
attributes are parameters and buffers. The hooks that are currently used
to automatically check serialization of every function and module in the
unit tests check that the archive produced by saving and loading and
saving again are the same and that the type tags for the actual IValues
representing the module match before saving and after loading. However,
these tests do not check that buffer and parameter metadata was not
lost or destroyed during serialization.

**Test Plan**
Ran the new unit test.

Differential Revision: [D25730603](https://our.internmc.facebook.com/intern/diff/D25730603)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add buffer/parameter metadata test to test_save_load.py"


**Summary**
This commit adds a unit test to `test_save_load.py` that checks that
saving and loading a module preserves metadata about which module
attributes are parameters and buffers. The hooks that are currently used
to automatically check serialization of every function and module in the
unit tests check that the archive produced by saving and loading and
saving again are the same and that the type tags for the actual IValues
representing the module match before saving and after loading. However,
these tests do not check that buffer and parameter metadata was not
lost or destroyed during serialization.

**Test Plan**
Ran the new unit test.

Differential Revision: [D25730603](https://our.internmc.facebook.com/intern/diff/D25730603)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add buffer/parameter metadata test to test_save_load.py"


**Summary**
This commit adds a unit test to `test_save_load.py` that checks that
saving and loading a module preserves metadata about which module
attributes are parameters and buffers. The hooks that are currently used
to automatically check serialization of every function and module in the
unit tests check that the archive produced by saving and loading and
saving again are the same and that the type tags for the actual IValues
representing the module match before saving and after loading. However,
these tests do not check that buffer and parameter metadata was not
lost or destroyed during serialization.

**Test Plan**
Ran the new unit test.

Differential Revision: [D25730603](https://our.internmc.facebook.com/intern/diff/D25730603)

[ghstack-poisoned]
